### ISSUE TITLE
Search for optimizations in the AI processing speed

### ISSUE THREAD

**Issue by abmpicoli2** [We-the-People-civ4col-mod/Mod#5]
@Nightinggale comment:
I figure the time spent by the AI is the time we should consider when optimizing. In most cases the game is fast enough during your turn and the slow code like getting yield production from a plot with modifiers is also used by the AI

@ShadesOT  comment:
That is something I noticed while studying the code
there are functions of magnitude n^2
or worse
by n^2 I mean o (n^2)
well, maybe it is too early to say something like o(n^2), i did not study the code there in depth.

**Comment by ShadesOT** [comment]
;-D

**Comment by ErikSandbraaten** [comment]
This is a big topic and I have a fairly comprehensive list of things to improve, what we could borrow / steal from other mods (K-Mod, C2C) and algorithms that could be entirely rewritten.
I also have a private build that used Intel TBB (A task parallelization library) which is similar to Java's fork-join and I parallelized 4 or 5 subsystems and achieved pretty decent performance gains.

Since the scope is so big, I think we need to find a way to discuss them in a more structured matter so that we don't get overwhelmed. I know that @Nightinggale is a "speed freak" and loves optimizing things as well, so I'm sure he's brimming with ideas too :D

**Comment by ErikSandbraaten** [comment]
I recently spent some investigating the Rise of Mankind: A New Dawn 2 DLL. It is very similar to C2C (one of them is technically a fork of the other...)

However, while profiling I noticed how efficient the AI pathfinding seemed to be. On huge maps, even in the late game it seemed to only account for about 20% or so of the CPU usage! This is quite impressive to say the least :)

After digging a bit deeper, I noticed that Koshling (the guru programmer of C2C) had implemented a kind of cache that the unit AI could query prior to making the actual pathfinding call.

The cache I am talking about is the class CvReachablePlotSet. The way it works is that rather than looping though all the plots in a given range, and then attempting to find a path there (filtered and scored by various criteria), a cache is built by visiting all the adjacent plots in a "ring-like" fashion. The first ring is 8 plots, the next is 16 and so on (it grows just like the city's outer radius / fat cross, albeit with corners included). 

It turns out that by querying this cache, a large amount of pathfinding calls can simply be eliminated \ culled. I think is reasonable to assume that merging this into WTP would lessen the pressure on the pathfinder significantly. Of course it would be even better if we also merge the K-Mod pathfinder (I am not convinced that Koshling's pathfinder is better, besides it also seems to have some bugs since I am getting PF asserts when running AND2)

We can actually do even better than this! 

The current implementation of CvReachablePlotSet is not even fully optimized:

- It uses recursion, but there is no reason why it should not be straight iteration. Recursion just makes it harder for the optimizer to eliminate the (stack) frame creation and it prevent proper inlining of the function body.
- It stores the plots it has visited in a hashmap and in the inner loop it checks if a new plot should be added by searching the hashmap (in the profiler I can see that a binary search is performed). This is redundant since we should be able to cull plots based on the ring's X and Y coordinate cutoff.
- The plot search order does not seem to be particularly cache efficient since plotDirection is used and the plots returned are not ordered in an ascending fashion.   
- If the ring size is below a certain margin, it may be better to just use a std::vector or even a fixed array.
The search can be implemented with SSE integer intrinsics which will probably demolish the hashmap in terms of performance due to cache friendliness unless the set becomes too large (and unless we write a custom allocator for the hashmap)
